lib/rouge/lexers/xpath.rb

```diff
@@ -0,0 +1,298 @@
+# -*- coding: utf-8 -*- #
+# frozen_string_literal: true
+
+module Rouge
+  module Lexers
+    class XPath < RegexLexer
+      title 'XPath'
+      desc 'XML Path Language (XPath) 3.1'
+      tag 'xpath'
+      filenames '*.xpath'
+
+      # Terminal literals:
+      # https://www.w3.org/TR/xpath-31/#terminal-symbols
+
+      digits = /[0-9]+/
+      decimalLiteral = /\.#{digits}|#{digits}\.[0-9]*/
+      doubleLiteral = /(\.#{digits})|#{digits}(\.[0-9]*)?[eE][+-]?#{digits}/
+      stringLiteral = /("(("")|[^"])*")|('(('')|[^'])*')/
+
+      ncName = /[a-z_][a-z_\-.0-9]*/i
+      qName = /(?:#{ncName})(?::#{ncName})?/
+      uriQName = /Q{[^{}]*}#{ncName}/
+      eqName = /(?:#{uriQName}|#{qName})/
+
+      commentStart = /\(:/
+      openParen    = /\((?!:)/
+
+      # Terminal symbols:
+      # https://www.w3.org/TR/xpath-30/#id-terminal-delimitation
+
+      kindTest = Regexp.union(%w(
+        element attribute schema-element schema-attribute
+        comment text node document-node namespace-node
+      ))
+      kindTestForPi = Regexp.union(%w(processing-instruction))
+      axes = Regexp.union(%w(
+        child descendant attribute self descendant-or-self
+        following-sibling following namespace
+        parent ancestor preceding-sibling preceding ancestor-or-self
+      ))
+      operators = Regexp.union(%w(, => = := : >= >> > <= << < - * != + // / || |))
+      keywords = Regexp.union(%w(let for some every if then else return in satisfies))
+      word_operators = Regexp.union(%w(
+        and or eq ge gt le lt ne is
+        div mod idiv
+        intersect except union
+        to
+      ))
+      constructorTypes = Regexp.union(%w(function array map empty-sequence))
+
+      # Lexical states:
+      # https://www.w3.org/TR/xquery-xpath-parsing/#XPath-lexical-states
+      # https://lists.w3.org/Archives/Public/public-qt-comments/2004Aug/0127.html
+      # https://www.w3.org/TR/xpath-30/#id-revision-log
+      # https://www.w3.org/TR/xpath-31/#id-revision-log
+
+      state :root do
+        # Comments
+        rule commentStart, Comment, :comment
```

Comment:
I'm actually having a little difficulty with this. If I define `commentStart` as an object method (`def self.commentStart`) as discussed above, I seem to have three ways of calling it:

1. `rule self.class.commentStart`, but it seems that the state block is evaluated in the context of a `Rouge::RegexLexer::StateDSL` instance, because that's what `self.class` refers to. Indeed, if I use this, I immediately get `undefined method 'commentStart' for Rouge::RegexLexer::StateDSL:Class`.
2. `rule XPath.commentStart` works, so that's what I've done for now
3. `rule commentStart` shouldn't work if I understand things correctly, but the (newly added) [guide](https://www.rubydoc.info/github/rouge-ruby/rouge/file/docs/LexerDevelopment.md#Special_Words) seems to say this works. If I try it, I immediately get `undefined local variable or method 'commentStart'`

Reply:
Sorry—I've got to run but need to think more about this. I'm pretty sure I've screwed up; both in my explanation to you and in the guide (so that'll need to be updated too). I think the way that you've done it is the right way to go about things but I just want to think more about it a bit later tonight (time permitting). Leaving this comment here as a placeholder.

Reply:
OK, me again. So:

1. You're absolutely correct about the respective calling contexts. I was forgetting that while the block passed to `rule` are evaluated in the context of an instance of the relevant lexer, the block passed to `state` is not. Given that, the way to refer to the method is by expressly referring to, in this case, the `XPath` singleton.

   I'm thinking it would really be a better recommendation overall to refer to memoised methods in this way, regardless of the evaluation context. The two things I worry about are brittleness and breaking polymorphism.

   The brittleness issue is, I think, real but more of a theoretical concern. Lexers don't really get renamed and rules are generally so short that I don't think it would be difficult to spot a mistaken reference.

   Whether this would break polymorphism I'm less sure about. My worry is that you'd have cross references that should 'just work' for subclassed lexers that reimplement some of these methods but that now won't because the reference is hard-coded to refer to the superclass (in this case, `XPath`). I think, though, that similar to the above this is more of a theoretical concern. I need to discuss with @jneen whether an approach should be standardised; I've been recommending `self.class` but perhaps that's been a mistake.
 
2. Looks good to me!

3. Ooh, well spotted. That was a careless mistake on my part.